CARDS-1374_demo

IntegralProgrammer opened the pull request:
This draft PR contains the work done thus far on the _CARDS-1374_ task. Right now, the following cases are handled:

- The user logs in with a username containing no `@` symbols. They are then prompted to provide their password for their local Apache Sling account.
- The user logs in with a username of `<USERNAME>@uhn.ca` and a browser-native popup window opens to a UHN Federated Services login page where they can provide their UHN password. _Please note that this login page will not actually log the user into CARDS as SAML authentication for CARDS has not yet been setup. The example URL that opens the UHN Federated Services login page was taken from the Office 365 login._
- The user attempts to login with a username containing an identity provider name other than `uhn.ca` or they specify multiple `@` symbols and a login error is displayed.

This behavior should happen both on the initial login screen as well as on any Material-UI login dialogs invoked by a call to `fetchWithReLogin`.

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Basic two-step login process setup

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

If the user has specified a remote domain (with the @ symbol), then check if the remote domain is listed under /apps/cards/SAMLDomains in the JCR repository. If it is not, display a failed login error

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

If the user has specified a valid remote domain, open a browser-native popup window to that domain's login page

IntegralProgrammer committed:
Add demo for using https://fs.uhn.ca as a SAML identity provider

marta- commented:
A couple of early suggestions:
* enable SAML authentication only in a specific runmode (e.g. add a `saml` runmode)
* turn on the 2-step login only if a SAML identity provider is configured, otherwise keep it in one step as before
* the "Username" label should become "Username or email address" if a SAML identity provider is configured
* update the error message when the user enters an unrecognized or malformed email address; right now it says "Invalid username or password"; it should say "Unrecognized email domain" or "Invalid email address" respectively
* focus the `password` input when it's displayed, and display the selected username in the label: "Password for Admin" instead of "Password"
* add a `Back` button on the password screen, in case the user realizes they misspelled their username
* move the popup to the center of the screen

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Auto-focus the Password input when it is displayed and display the selected username in the label

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Fix indentation

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Add a Back button on the password entry phase so that the user can go backwards if they mistyped their username

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Cause the popup window for the identity provider login to open in the center of the screen

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Display more meaningful error messages

IntegralProgrammer committed:
CARDS-1374: Convert SignIn (loginForm.js) component to a two-step process - one screen for username entry, one screen for password entry

Only operate in two-step mode if the /apps/cards/SAMLDomains JCR directory exists

IntegralProgrammer committed:
Move cards-modules-uhn-saml artifact to its own saml runMode

IntegralProgrammer commented:
>     * enable SAML authentication only in a specific runmode (e.g. add a `saml` runmode)

I have addressed this by only loading the `io.uhndata.cards/cards-modules-uhn-saml/${cards.version}` artifact (which adds a JCR entry for SAML authentication with UHN under `/apps/cards/SAMLDomains`) when the `saml` runMode is enabled.

IntegralProgrammer commented:
>     * turn on the 2-step login on only if a SAML identity provider is configured, otherwise keep it in one step as before

I have addressed this by instructing the `SignIn` component to check if the `/apps/cards/SAMLDomains` JCR folder is present. If it is present, then SAML is available and the `SignIn` component behaves in two-step mode. Otherwise, the `SignIn` component behaves in one-step mode.

IntegralProgrammer commented:
>     * the "Username" label should become "Username or email address" if a SAML identity provider is configured

The _Username_ label now switches to _Username or email address_ if the `SignIn` component is operating in two-step mode.

IntegralProgrammer commented:
>     * update the error message when the user enters an unrecognized or malformed email address; right now it says "Invalid username or password"; it should say "Unrecognized email domain" or "Invalid email address" respectively
> 
>     * focus the `password` input when it's displayed, and display the selected username in the label: "Password for Admin" instead of "Password"
> 
>     * add a `Back` button on the password screen, in case the user realizes they misspelled their username
> 
>     * move the popup to the center of the screen

Done :heavy_check_mark:

IntegralProgrammer commented:
>     * `fetchWithReLogin` should pop the remote authentication page if the current SAML is enabled and the current user is not local

This seems somewhat out-of-scope for this PR. I think it would be nice, regardless of how the user logged in (eg. local Sling user, LDAP, SAML, etc...) for the `fetchWithReLogin` dialog to automatically be populated with the username associated with the expired session. It looks like this has already been discussed in https://github.com/data-team-uhn/cards/blob/1bfb49740b7c03db72464a882d760877dc9da27b/modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx#L63 which, is required to be implemented so as to be able to implement the _username auto-population_ feature.

marta- commented:
Closing this outdated demo PR in favor of #815.
